README.md
# flutter_ripple

A simple flutter package to show rippple effect in any widget 

## Sample Image

![alt text](https://github.com/Destiny-Ed/flutter_ripple/blob/main/assets/ripple.gif)

## Usage

[Example] (https://github.com/Destiny-Ed/flutter_ripple/blob/main/example/example.dart)

To use this package : *add the dependency to your [pubspec.yaml] file

```yaml
   dependencies:
        flutter:
            sdk : flutter
        flutter_ripple: ^0.0.1
```

## Add to your dart file

```dart
import 'package:flutter/material.dart';

import 'package:flutter_ripple/flutter_ripple.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: RippleExample(),
    );
  }
}



class RippleExample extends StatefulWidget {
  const RippleExample({Key? key}) : super(key: key);

  @override
  _RippleExampleState createState() => _RippleExampleState();
}

class _RippleExampleState extends State<RippleExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter Ripple'),
      ),
      body: Center(
        child: Stack(
          children: [
            ///Show Ripple Widget
            FlutterRipple(
              child: Text("Flutter Ripple"),
              rippleColor: Colors.blue,
              onTap: () {
                print("hello");
              },
            ),
            Positioned(
              top: MediaQuery.of(context).size.height / 2 + 100,
              right: 50,
              left: 50,
              child: Container(
                child: Column(
                  children: [
                    Text(
                      "Searching for location.",
                      textAlign: TextAlign.center,
                    ),
                    Container(
                      margin: const EdgeInsets.only(top: 10),
                      padding: const EdgeInsets.symmetric(
                          horizontal: 30, vertical: 15),
                      decoration: BoxDecoration(
                          color: Colors.blue,
                          borderRadius: BorderRadius.circular(50)),
                      child: Text(
                        "Find All Location",
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}

```

## Say Hi To Me

- Twitter : @DestinyEd8
- Linkedin : www.linkedin.com/in/destiny-ed

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.1] - July 18 2021.

* First Release

## [0.0.2]

* Users can specify the ripple radius

## [0.0.3]

* Fixed 'animationController not dispose' properly thereby avoiding memory leak



test/flutter_ripple_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_ripple/flutter_ripple.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/flutter_ripple.dart
library flutter_ripple;

import 'package:flutter/material.dart';

///A Flutter Class For Showing Ripple Effects
class FlutterRipple extends StatefulWidget {
  FlutterRipple(
      {required this.child,
      Key? key,
      this.rippleColor,
      this.duration,
      this.onTap,
      this.rippleShape,
      this.radius})
      : super(key: key);

  final Duration? duration;
  final double? radius;
  final Widget? child;
  final Color? rippleColor;
  final BoxShape? rippleShape;
  final VoidCallback? onTap;

  @override
  _FlutterRippleState createState() => _FlutterRippleState();
}

///A Simple Widget that show how the flutter ripple package is used

/// ```dart

///   @override
///   Widget build(BuildContext context) {
///     return Scaffold(
///       appBar: AppBar(
///         title: const Text('Flutter Ripple'),
///       ),
///       body: Center(
///     children: [
///      FlutterRipple(
///         child: Text("Flutter Ripple"),
///         rippleColor: MyColors().blue,
///         onTap: () {
///           print("hello");
///         },
///       ),
///       Positioned(
///           top: MediaQuery.of(context).size.height / 2 + 100,
///           right: 50,
///           left: 50,
///           child: Container(
///             child: Column(
///               children: [
///                 Text(
///                   "Searching for location.",
///                   textAlign: TextAlign.center,
///                 ),
///                 Container(
///                  margin: const EdgeInsets.only(top: 10),
///                   padding: const EdgeInsets.symmetric(
///                       horizontal: 30, vertical: 15),
///                   decoration: BoxDecoration(
///                       color: Colors.blue,
///                       borderRadius: BorderRadius.circular(50)),
///                   child: Text(
///                     "Find All Location",
///                     style: TextStyle(color: Colors.white),
///                   ),
///                 )
///              ],
///             ),
///           ))
///     ],
///   ),
/// );
///     );
///
/// }
/// ```

class _FlutterRippleState extends State<FlutterRipple>
    with SingleTickerProviderStateMixin {
  AnimationController? _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      lowerBound: 0.5,
      duration: widget.duration ?? Duration(seconds: 2),
    )..repeat();
  }

  @override
  void dispose() {
    _controller!.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(onTap: widget.onTap ?? null, child: _bodyWidget());
  }

  Widget _bodyWidget() {
    return AnimatedBuilder(
      animation:
          CurvedAnimation(parent: _controller!, curve: Curves.fastOutSlowIn),
      builder: (context, child) {
        return Stack(
          alignment: Alignment.center,
          children: <Widget>[
            ...List.generate(
              4,
              (index) => _containerWidget(widget.radius == null
                  ? (100 * index) * _controller!.value
                  : (widget.radius! * index) * _controller!.value),
            ),
            Align(child: widget.child),
          ],
        );
      },
    );
  }

  Widget _containerWidget(double radius) {
    return Container(
      width: radius,
      height: radius,
      decoration: BoxDecoration(
          shape: widget.rippleShape ?? BoxShape.circle,
          color: widget.rippleColor == null
              ? Colors.blue.withOpacity(1 - _controller!.value)
              : widget.rippleColor!.withOpacity(1 - _controller!.value)),
    );
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:flutter_ripple/flutter_ripple.dart';

void main() {
  runApp(App());
}

class App extends StatelessWidget {
  const App({ Key? key }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(

      home: RippleExample(),
      
    );
  }
}
class RippleExample extends StatefulWidget {
  const RippleExample({Key? key}) : super(key: key);

  @override
  _RippleExampleState createState() => _RippleExampleState();
}

class _RippleExampleState extends State<RippleExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Flutter Ripple'),
      ),
      body: Center(
        child: Stack(
          children: [
            ///Show Ripple Widget
            FlutterRipple(
              child: Text("Flutter Ripple"),
              rippleColor: Colors.blue,
              onTap: () {
                print("hello");
              },
            ),
            Positioned(
              top: MediaQuery.of(context).size.height / 2 + 100,
              right: 50,
              left: 50,
              child: Container(
                child: Column(
                  children: [
                    Text(
                      "Searching for location.",
                      textAlign: TextAlign.center,
                    ),
                    Container(
                      margin: const EdgeInsets.only(top: 10),
                      padding: const EdgeInsets.symmetric(
                          horizontal: 30, vertical: 15),
                      decoration: BoxDecoration(
                          color: Colors.blue,
                          borderRadius: BorderRadius.circular(50)),
                      child: Text(
                        "Find All Location",
                        style: TextStyle(color: Colors.white),
                      ),
                    )
                  ],
                ),
              ),
            )
          ],
        ),
      ),
    );
  }
}
